src/plugins/dist_matrixops/retile_annotations.cpp

```diff
@@ -0,0 +1,811 @@
+    execution_tree::match_pattern_type const retile_annotations::match_data =
+    {
+        hpx::util::make_tuple("retile_d", std::vector<std::string>{R"(
+                retile_d(
+                    _1_a,
+                    __arg(_2_tiling_type, "sym"),
+                    __arg(_3_intersection, nil),
+                    __arg(_4_numtiles, num_localities()),
+                    __arg(_5_new_tiling, nil)
+                )
+            )"},
+            &create_retile_annotations,
+            &execution_tree::create_primitive<retile_annotations>, R"(
+            a, tiling_type, intersection, numtiles, new_tiling
+            Args:
+
+                a (array): a distributed array. A vector or a matrix.
+                tiling_type (string, optional): defaults to `sym` which is a
+                    balanced way of tiling among all localities. Other options
+                    are `row`, `column` or `user` tiling types. If an
+                    intersection is given, retile produces tiles that have
+                    overlapped parts. In the `user` mode, using new_tiling,
+                    tiles are specified with their spans.
+                intersection (int or tuple of ints, optional): the size of
+                    overlapped part on each dimension. If an integer is given,
+                    that would be the intersection length on all dimensions
+                    that are tiled. If the given intersection is odd, the extra
+                    overlapped vector will be at the end of the part.
+                    In the `user` mode, `intersection` cannot be used.
+                numtiles (int, optional): number of tiles of the returned array
+                    If not given it sets to the number of localities in the
+                    application.
+                new_tiling (list, optional): a new tiling specification for the
+                    current locality that is specified only for the `user` mode
+                    E.g. for a matrix we can have:
+                    list("tile", list("columns", 0, 2), list("rows", 0, 2)) or
+                    list("args", list("locality", 0, 4),
+                       list("tile", list("columns", 0, 2), list("rows", 0, 2)))
+
+            Returns:
+
+            A retiled array according to the tiling type on numtiles localities
+            The returned array can contain overlapped parts)")
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    retile_annotations::retile_annotations(
+        execution_tree::primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    ///////////////////////////////////////////////////////////////////////////
+    namespace detail
+    {
+        std::tuple<bool, std::int64_t, std::int64_t> tile_extraction_1d_helper(
+            ir::range_iterator&& it)
```

Comment:
Does this have to be an rvalue? I think passing iterators by value is normally safe.

Reply:
Okay, sure